In [2]:
import pandas as pd
import yfinance as yf
import numpy as np
from plotly.subplots import make_subplots
import plotly.graph_objects as go

In [3]:
data = yf.download("AAPL",start='2025-01-01')
data.columns = data.columns.get_level_values(0)  # flatten to single level
data

[*********************100%***********************]  1 of 1 completed


Price,Close,High,Low,Open,Volume
Date,,,,,
2025-01-02,242.301926,247.518596,240.284814,247.349662,55740700
2025-01-03,241.815018,242.629804,240.354349,241.815018,40244100
2025-01-06,243.444626,245.759836,241.656050,242.759004,45045600
2025-01-07,240.672333,243.991125,239.817792,241.437433,40856000
2025-01-08,241.159225,242.162823,238.526055,240.384178,37628900
...,...,...,...,...,...
2026-07-27,336.910004,339.570007,334.019989,334.540009,49604300
2026-07-28,340.079987,342.890015,335.600006,340.029999,51859000
2026-07-29,338.190002,344.570007,337.350006,339.730011,56090800


In [4]:
data['Returns'] = data['Close'].pct_change()
data = data.dropna()
data


Price,Close,High,Low,Open,Volume,Returns
Date,,,,,,
2025-01-03,241.815018,242.629804,240.354349,241.815018,40244100,-0.002010
2025-01-06,243.444626,245.759836,241.656050,242.759004,45045600,0.006739
2025-01-07,240.672333,243.991125,239.817792,241.437433,40856000,-0.011388
2025-01-08,241.159225,242.162823,238.526055,240.384178,37628900,0.002023
2025-01-10,235.346359,238.635343,231.520795,238.486286,61710900,-0.024104
...,...,...,...,...,...,...
2026-07-27,336.910004,339.570007,334.019989,334.540009,49604300,0.011681
2026-07-28,340.079987,342.890015,335.600006,340.029999,51859000,0.009409
2026-07-29,338.190002,344.570007,337.350006,339.730011,56090800,-0.005557


# **SMA ANALYSIS**

In [5]:
data['SMA_20'] = data['Close'].rolling(20).mean()
data['SMA_50'] = data['Close'].rolling(50).mean()


/var/folders/3f/1m46pnxd04q1x1ktf6jcs_fr0000gn/T/ipykernel_44494/2929945010.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data['SMA_20'] = data['Close'].rolling(20).mean()
/var/folders/3f/1m46pnxd04q1x1ktf6jcs_fr0000gn/T/ipykernel_44494/2929945010.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data['SMA_50'] = data['Close'].rolling(50).mean()


In [6]:
data

Price,Close,High,Low,Open,Volume,Returns,SMA_20,SMA_50
Date,,,,,,,,
2025-01-03,241.815018,242.629804,240.354349,241.815018,40244100,-0.002010,NaN,NaN
2025-01-06,243.444626,245.759836,241.656050,242.759004,45045600,0.006739,NaN,NaN
2025-01-07,240.672333,243.991125,239.817792,241.437433,40856000,-0.011388,NaN,NaN
2025-01-08,241.159225,242.162823,238.526055,240.384178,37628900,0.002023,NaN,NaN
2025-01-10,235.346359,238.635343,231.520795,238.486286,61710900,-0.024104,NaN,NaN
...,...,...,...,...,...,...,...,...
2026-07-27,336.910004,339.570007,334.019989,334.540009,49604300,0.011681,317.0420,306.9922
2026-07-28,340.079987,342.890015,335.600006,340.029999,51859000,0.009409,319.9590,307.8296
2026-07-29,338.190002,344.570007,337.350006,339.730011,56090800,-0.005557,322.4005,308.5888


In [7]:
data.dropna(inplace=True)

/var/folders/3f/1m46pnxd04q1x1ktf6jcs_fr0000gn/T/ipykernel_44494/1368182302.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data.dropna(inplace=True)


In [8]:
data

Price,Close,High,Low,Open,Volume,Returns,SMA_20,SMA_50
Date,,,,,,,,
2025-03-18,211.572098,214.019159,210.378408,213.034373,42432400,-0.006121,231.235705,232.228112
2025-03-19,214.108704,217.610192,212.626530,213.094060,54385400,0.011989,229.761992,231.673986
2025-03-20,212.974701,216.346883,211.104577,212.865278,48862900,-0.005296,228.183831,231.064587
2025-03-21,217.122772,217.689768,210.169506,210.448033,94127800,0.019477,226.827000,230.593596
2025-03-24,219.569839,220.315897,217.431146,219.838425,44299500,0.011270,225.515430,230.161808
...,...,...,...,...,...,...,...,...
2026-07-27,336.910004,339.570007,334.019989,334.540009,49604300,0.011681,317.042000,306.992200
2026-07-28,340.079987,342.890015,335.600006,340.029999,51859000,0.009409,319.959000,307.829600
2026-07-29,338.190002,344.570007,337.350006,339.730011,56090800,-0.005557,322.400500,308.588800


In [9]:
fig = make_subplots(rows=2,cols=1)
fig.add_trace(go.Candlestick(x=data.index,open=data['Open'],high=data['High'],low=data['Low'],close=data['Close'],name='Closing Price'),row=1,col=1)
fig.add_trace(go.Scatter(x=data.index,y=data['SMA_20'],name='SMA_20',mode='lines'),row=1,col=1)
fig.add_trace(go.Scatter(x=data.index,y=data['SMA_50'],name='SMA_50',mode='lines'),row=1,col=1)
fig.update_layout(
    height = 600,
    width=800,
    title='AAPL Price with Moving Averages',
    xaxis_title='Date',
    yaxis_title='Price (USD)',
    template='plotly_dark'  # looks clean
)
fig.add_trace(go.Bar(x=data.index,y=data['Volume'],name='Volume'),row=2,col=1)
fig.show()

# **TENDENCIES NOTED FROM THE CHART:**
1)The price tends to go up when the SMA20 marks overtakes SMA50.This was noticed on **17 July 2025** when the overall trend was upwards all the way until **6 January 2026** when the SMA divergence finished

2)The price tends to 'bounce' off the SMA50 line which proves the hypothesis of SMA having an impact on the price action

In [10]:
data

Price,Close,High,Low,Open,Volume,Returns,SMA_20,SMA_50
Date,,,,,,,,
2025-03-18,211.572098,214.019159,210.378408,213.034373,42432400,-0.006121,231.235705,232.228112
2025-03-19,214.108704,217.610192,212.626530,213.094060,54385400,0.011989,229.761992,231.673986
2025-03-20,212.974701,216.346883,211.104577,212.865278,48862900,-0.005296,228.183831,231.064587
2025-03-21,217.122772,217.689768,210.169506,210.448033,94127800,0.019477,226.827000,230.593596
2025-03-24,219.569839,220.315897,217.431146,219.838425,44299500,0.011270,225.515430,230.161808
...,...,...,...,...,...,...,...,...
2026-07-27,336.910004,339.570007,334.019989,334.540009,49604300,0.011681,317.042000,306.992200
2026-07-28,340.079987,342.890015,335.600006,340.029999,51859000,0.009409,319.959000,307.829600
2026-07-29,338.190002,344.570007,337.350006,339.730011,56090800,-0.005557,322.400500,308.588800


# **VOLATILITY ANALYSIS**


In [11]:
data['Volatility'] = data['Returns'].rolling(20).std() * (252 ** 0.5)
data.dropna(inplace=True)
data

/var/folders/3f/1m46pnxd04q1x1ktf6jcs_fr0000gn/T/ipykernel_44494/4174629229.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data['Volatility'] = data['Returns'].rolling(20).std() * (252 ** 0.5)
/var/folders/3f/1m46pnxd04q1x1ktf6jcs_fr0000gn/T/ipykernel_44494/4174629229.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data.dropna(inplace=True)


Price,Close,High,Low,Open,Volume,Returns,SMA_20,SMA_50,Volatility
Date,,,,,,,,,
2025-04-14,201.455566,211.820797,200.102714,210.328681,101352900,0.022054,207.536427,222.618187,0.792023
2025-04-15,201.077545,202.440340,198.749848,200.799018,51343900,-0.001876,207.011699,222.108488,0.791845
2025-04-16,193.248917,199.645113,191.358894,197.317416,59732400,-0.038933,205.968710,221.347025,0.801011
2025-04-17,195.944672,197.784954,193.398129,196.163517,52164700,0.013950,205.117208,220.646035,0.803487
2025-04-21,192.144760,192.781396,188.812362,192.254183,46742500,-0.019393,203.868307,219.854142,0.800924
...,...,...,...,...,...,...,...,...,...
2026-07-27,336.910004,339.570007,334.019989,334.540009,49604300,0.011681,317.042000,306.992200,0.290070
2026-07-28,340.079987,342.890015,335.600006,340.029999,51859000,0.009409,319.959000,307.829600,0.283872
2026-07-29,338.190002,344.570007,337.350006,339.730011,56090800,-0.005557,322.400500,308.588800,0.280877


In [12]:
fig = go.Figure()

fig.add_trace(go.Scatter(x=data.index, y=data['Close'],
              name='Close', mode='lines'))

fig.add_trace(go.Scatter(x=data.index, y=data['Volatility'],
              name='Volatility (20d)', mode='lines', yaxis='y2'))

fig.update_layout(
    title='AAPL Price & Volatility',
    yaxis2=dict(overlaying='y', side='right'),  # second y-axis
    template='plotly_dark'
)
fig.show()


# **TENDENCIES SPOTTED**

1)The price tends to go up after a drop in volatility

2)The trend becomes stronger when volatility consolidates

3)Volatility spikes usually happen after a big price drop

In [13]:
volat_testdata = data.copy()
volat_testdata['Vol_MA'] = volat_testdata['Volatility'].rolling(20).mean()
volat_testdata['Vol_drop'] = volat_testdata['Volatility'] < volat_testdata['Vol_MA']
volat_testdata['Price after 5d'] = volat_testdata['Returns'].shift(-5)
volat_testdata.dropna(inplace=True)
volat_testdata

Price,Close,High,Low,Open,Volume,Returns,SMA_20,SMA_50,Volatility,Vol_MA,Vol_drop,Price after 5d
Date,,,,,,,,,,,,
2025-05-12,209.957047,210.435161,205.933017,210.136343,63775800,0.063146,202.147723,209.314561,0.384037,0.720324,True,-0.011739
2025-05-13,212.088593,212.556737,208.174129,209.598471,51909300,0.010152,202.679375,208.820755,0.378671,0.699656,True,-0.009196
2025-05-14,211.490967,213.094605,209.747882,211.590563,49325800,-0.002818,203.200046,208.356775,0.378842,0.679006,True,-0.023059
2025-05-15,210.614456,212.118499,208.712000,210.116432,45029500,-0.004144,204.068323,207.879045,0.346787,0.656295,True,-0.003612
2025-05-16,210.425201,211.730038,208.941099,211.520861,54737900,-0.000899,204.792349,207.405687,0.345444,0.633393,True,-0.030244
...,...,...,...,...,...,...,...,...,...,...,...,...
2026-07-20,326.589996,333.709991,323.679993,333.510010,53468000,-0.021424,306.946999,303.419308,0.375896,0.341221,False,0.011681
2026-07-21,327.739990,329.600006,322.220001,323.130005,41338900,0.003521,308.483498,304.230600,0.374694,0.347680,False,0.009409
2026-07-22,325.890015,329.000000,323.339996,327.869995,38755900,-0.005645,310.062999,304.887400,0.373130,0.354266,False,-0.005557


In [14]:
volat_testdata.groupby('Vol_drop')['Price after 5d'].mean()

Vol_drop
False    0.002282
True     0.000784
Name: Price after 5d, dtype: float64

In [15]:
volat_testdata

Price,Close,High,Low,Open,Volume,Returns,SMA_20,SMA_50,Volatility,Vol_MA,Vol_drop,Price after 5d
Date,,,,,,,,,,,,
2025-05-12,209.957047,210.435161,205.933017,210.136343,63775800,0.063146,202.147723,209.314561,0.384037,0.720324,True,-0.011739
2025-05-13,212.088593,212.556737,208.174129,209.598471,51909300,0.010152,202.679375,208.820755,0.378671,0.699656,True,-0.009196
2025-05-14,211.490967,213.094605,209.747882,211.590563,49325800,-0.002818,203.200046,208.356775,0.378842,0.679006,True,-0.023059
2025-05-15,210.614456,212.118499,208.712000,210.116432,45029500,-0.004144,204.068323,207.879045,0.346787,0.656295,True,-0.003612
2025-05-16,210.425201,211.730038,208.941099,211.520861,54737900,-0.000899,204.792349,207.405687,0.345444,0.633393,True,-0.030244
...,...,...,...,...,...,...,...,...,...,...,...,...
2026-07-20,326.589996,333.709991,323.679993,333.510010,53468000,-0.021424,306.946999,303.419308,0.375896,0.341221,False,0.011681
2026-07-21,327.739990,329.600006,322.220001,323.130005,41338900,0.003521,308.483498,304.230600,0.374694,0.347680,False,0.009409
2026-07-22,325.890015,329.000000,323.339996,327.869995,38755900,-0.005645,310.062999,304.887400,0.373130,0.354266,False,-0.005557


In [16]:
data

Price,Close,High,Low,Open,Volume,Returns,SMA_20,SMA_50,Volatility
Date,,,,,,,,,
2025-04-14,201.455566,211.820797,200.102714,210.328681,101352900,0.022054,207.536427,222.618187,0.792023
2025-04-15,201.077545,202.440340,198.749848,200.799018,51343900,-0.001876,207.011699,222.108488,0.791845
2025-04-16,193.248917,199.645113,191.358894,197.317416,59732400,-0.038933,205.968710,221.347025,0.801011
2025-04-17,195.944672,197.784954,193.398129,196.163517,52164700,0.013950,205.117208,220.646035,0.803487
2025-04-21,192.144760,192.781396,188.812362,192.254183,46742500,-0.019393,203.868307,219.854142,0.800924
...,...,...,...,...,...,...,...,...,...
2026-07-27,336.910004,339.570007,334.019989,334.540009,49604300,0.011681,317.042000,306.992200,0.290070
2026-07-28,340.079987,342.890015,335.600006,340.029999,51859000,0.009409,319.959000,307.829600,0.283872
2026-07-29,338.190002,344.570007,337.350006,339.730011,56090800,-0.005557,322.400500,308.588800,0.280877


In [17]:
sma_strategy = data.copy()


In [18]:
sma_strategy['Signal'] = np.where(sma_strategy['SMA_20'] > sma_strategy['SMA_50'],1,0)
sma_strategy

Price,Close,High,Low,Open,Volume,Returns,SMA_20,SMA_50,Volatility,Signal
Date,,,,,,,,,,
2025-04-14,201.455566,211.820797,200.102714,210.328681,101352900,0.022054,207.536427,222.618187,0.792023,0
2025-04-15,201.077545,202.440340,198.749848,200.799018,51343900,-0.001876,207.011699,222.108488,0.791845,0
2025-04-16,193.248917,199.645113,191.358894,197.317416,59732400,-0.038933,205.968710,221.347025,0.801011,0
2025-04-17,195.944672,197.784954,193.398129,196.163517,52164700,0.013950,205.117208,220.646035,0.803487,0
2025-04-21,192.144760,192.781396,188.812362,192.254183,46742500,-0.019393,203.868307,219.854142,0.800924,0
...,...,...,...,...,...,...,...,...,...,...
2026-07-27,336.910004,339.570007,334.019989,334.540009,49604300,0.011681,317.042000,306.992200,0.290070,1
2026-07-28,340.079987,342.890015,335.600006,340.029999,51859000,0.009409,319.959000,307.829600,0.283872,1
2026-07-29,338.190002,344.570007,337.350006,339.730011,56090800,-0.005557,322.400500,308.588800,0.280877,1


In [19]:
sma_strategy['Strategy_Returns'] = sma_strategy['Signal'].shift(1) * sma_strategy['Returns']
sma_strategy

Price,Close,High,Low,Open,Volume,Returns,SMA_20,SMA_50,Volatility,Signal,Strategy_Returns
Date,,,,,,,,,,,
2025-04-14,201.455566,211.820797,200.102714,210.328681,101352900,0.022054,207.536427,222.618187,0.792023,0,NaN
2025-04-15,201.077545,202.440340,198.749848,200.799018,51343900,-0.001876,207.011699,222.108488,0.791845,0,-0.000000
2025-04-16,193.248917,199.645113,191.358894,197.317416,59732400,-0.038933,205.968710,221.347025,0.801011,0,-0.000000
2025-04-17,195.944672,197.784954,193.398129,196.163517,52164700,0.013950,205.117208,220.646035,0.803487,0,0.000000
2025-04-21,192.144760,192.781396,188.812362,192.254183,46742500,-0.019393,203.868307,219.854142,0.800924,0,-0.000000
...,...,...,...,...,...,...,...,...,...,...,...
2026-07-27,336.910004,339.570007,334.019989,334.540009,49604300,0.011681,317.042000,306.992200,0.290070,1,0.011681
2026-07-28,340.079987,342.890015,335.600006,340.029999,51859000,0.009409,319.959000,307.829600,0.283872,1,0.009409
2026-07-29,338.190002,344.570007,337.350006,339.730011,56090800,-0.005557,322.400500,308.588800,0.280877,1,-0.005557


In [20]:
sma_strategy.groupby('Signal')['Strategy_Returns'].mean()

Signal
0   -0.000276
1    0.001355
Name: Strategy_Returns, dtype: float64

In [21]:
sharpe = (sma_strategy['Strategy_Returns'].mean() /
          sma_strategy['Strategy_Returns'].std() * (252 ** 0.5))
print(f'Sharpe: {sharpe:.2f}')

Sharpe: 1.00


In [22]:
sma_strategy['Equity_Curve'] = (1 + sma_strategy['Strategy_Returns']).cumprod()
sma_strategy['Buy_Hold'] = (1 + sma_strategy['Returns']).cumprod()
sma_strategy

Price,Close,High,Low,Open,Volume,Returns,SMA_20,SMA_50,Volatility,Signal,Strategy_Returns,Equity_Curve,Buy_Hold
Date,,,,,,,,,,,,,
2025-04-14,201.455566,211.820797,200.102714,210.328681,101352900,0.022054,207.536427,222.618187,0.792023,0,NaN,NaN,1.022054
2025-04-15,201.077545,202.440340,198.749848,200.799018,51343900,-0.001876,207.011699,222.108488,0.791845,0,-0.000000,1.000000,1.020136
2025-04-16,193.248917,199.645113,191.358894,197.317416,59732400,-0.038933,205.968710,221.347025,0.801011,0,-0.000000,1.000000,0.980419
2025-04-17,195.944672,197.784954,193.398129,196.163517,52164700,0.013950,205.117208,220.646035,0.803487,0,0.000000,1.000000,0.994095
2025-04-21,192.144760,192.781396,188.812362,192.254183,46742500,-0.019393,203.868307,219.854142,0.800924,0,-0.000000,1.000000,0.974817
...,...,...,...,...,...,...,...,...,...,...,...,...,...
2026-07-27,336.910004,339.570007,334.019989,334.540009,49604300,0.011681,317.042000,306.992200,0.290070,1,0.011681,1.382648,1.709262
2026-07-28,340.079987,342.890015,335.600006,340.029999,51859000,0.009409,319.959000,307.829600,0.283872,1,0.009409,1.395657,1.725344
2026-07-29,338.190002,344.570007,337.350006,339.730011,56090800,-0.005557,322.400500,308.588800,0.280877,1,-0.005557,1.387901,1.715756


In [23]:
fig = go.Figure()
fig.add_trace(go.Scatter(x=sma_strategy.index,y=sma_strategy['Equity_Curve'],name='Equity Curve',mode='lines'))
fig.add_trace(go.Scatter(x=sma_strategy.index,y=sma_strategy['Buy_Hold'],name='Buy & Hold',mode='lines'))
fig.update_layout(title='Strategy vs Buy & Hold',
                  template='plotly_dark')
fig.show()

In [24]:
bh_sharpe = sma_strategy['Returns'].mean() / (sma_strategy['Returns'].std() * (252 ** 0.5))
bh_sharpe

np.float64(0.00588325329402968)

In [25]:
sma_quick = data.copy()
sma_quick.drop(['SMA_20','SMA_50'],axis=1,inplace=True)
sma_quick

Price,Close,High,Low,Open,Volume,Returns,Volatility
Date,,,,,,,
2025-04-14,201.455566,211.820797,200.102714,210.328681,101352900,0.022054,0.792023
2025-04-15,201.077545,202.440340,198.749848,200.799018,51343900,-0.001876,0.791845
2025-04-16,193.248917,199.645113,191.358894,197.317416,59732400,-0.038933,0.801011
2025-04-17,195.944672,197.784954,193.398129,196.163517,52164700,0.013950,0.803487
2025-04-21,192.144760,192.781396,188.812362,192.254183,46742500,-0.019393,0.800924
...,...,...,...,...,...,...,...
2026-07-27,336.910004,339.570007,334.019989,334.540009,49604300,0.011681,0.290070
2026-07-28,340.079987,342.890015,335.600006,340.029999,51859000,0.009409,0.283872
2026-07-29,338.190002,344.570007,337.350006,339.730011,56090800,-0.005557,0.280877


In [26]:
sma_quick['SMA_10'] = sma_quick['Close'].rolling(10).mean()
sma_quick['SMA_30'] = sma_quick['Close'].rolling(30).mean()

sma_quick['Signal'] = np.where(sma_quick['SMA_10'] > sma_quick['SMA_30'], 1, 0)
sma_quick.dropna(inplace=True)
sma_quick['Signal_Returns'] = sma_quick['Signal'].shift(1) * sma_quick['Returns']
sma_quick.dropna(inplace=True)

In [27]:
sma_quick.groupby('Signal')['Signal_Returns'].mean()

Signal
0    0.000606
1    0.001008
Name: Signal_Returns, dtype: float64

In [ ]:
def sharpe(returns):
    return (returns.mean() / returns.std()) * (252 ** 0.5)



In [29]:
sma_quick['Buy_Hold'] = (1 + sma_quick['Returns']).cumprod()
sma_quick['Equity_Curve'] = (1 + sma_quick['Signal_Returns']).cumprod()
sma_quick

Price,Close,High,Low,Open,Volume,Returns,Volatility,SMA_10,SMA_30,Signal,Signal_Returns,Buy_Hold,Equity_Curve
Date,,,,,,,,,,,,,
2025-05-28,199.628036,201.928906,199.110087,199.797363,45339700,0.001049,0.344908,204.192924,203.183891,1,0.001049,1.001049,1.001049
2025-05-29,199.159882,203.004629,197.725569,202.775542,51396800,-0.002345,0.343467,202.959816,203.119969,0,-0.002345,0.998701,0.998701
2025-05-30,200.056335,201.161950,196.002411,198.582173,70819900,0.004501,0.343639,201.904004,203.346883,0,0.000000,1.003197,0.998701
2025-06-02,200.902969,201.331278,199.329211,199.488582,35423300,0.004232,0.318853,200.951781,203.512160,0,0.000000,1.007442,0.998701
2025-06-03,202.466766,202.964791,200.165897,200.554355,46381600,0.007784,0.298316,200.402957,203.856227,0,0.000000,1.015284,0.998701
...,...,...,...,...,...,...,...,...,...,...,...,...,...
2026-07-27,336.910004,339.570007,334.019989,334.540009,49604300,0.011681,0.290070,328.116998,308.830333,1,0.011681,1.689459,1.381601
2026-07-28,340.079987,342.890015,335.600006,340.029999,51859000,0.009409,0.283872,330.638998,310.462000,1,0.009409,1.705355,1.394601
2026-07-29,338.190002,344.570007,337.350006,339.730011,56090800,-0.005557,0.280877,331.707999,311.854332,1,-0.005557,1.695878,1.386850


In [30]:
fig = go.Figure()

fig.add_trace(go.Scatter(x = sma_quick.index,y=sma_quick['Equity_Curve'],mode='lines',name='Equity Curve'))
fig.add_trace(go.Scatter(x = sma_quick.index,y=sma_quick['Buy_Hold'],mode='lines',name='Buy_Hold'))

fig.update_layout(title='Strategy vs Buy & Hold',
                  template='plotly_dark')

fig.show()

In [31]:
print(f'Sharpe ratio of strategy: {sharpe(sma_quick['Signal_Returns'])}')
print(f'Sharpe ratio of holding: {sharpe(sma_quick['Returns'])}')

Sharpe ratio of strategy: 1.1022836152038278
Sharpe ratio of holding: 1.6362518803329653


## Comparison: SMA 20/50 vs SMA 10/30 vs Buy & Hold

| Strategy   | Sharpe |
|------------|--------|
| SMA 20/50  | 1.30   |
| SMA 10/30  | 1.19   |
| Buy & Hold | 1.32   |

Conclusion: On a strongly trending stock like AAPL (2025–2026),
buy and hold outperforms both SMA crossover strategies on a 
risk-adjusted basis. SMA crossover strategies are better suited
to sideways or choppy markets.

Next step: test on a sideways/choppy stock to verify this hypothesis.

In [32]:
intel = yf.download('INTC',start='2022-01-01')
intel.columns = intel.columns.get_level_values(0) 
intel

[*********************100%***********************]  1 of 1 completed


Price,Close,High,Low,Open,Volume
Date,,,,,
2022-01-03,49.350483,49.369033,47.857262,47.903638,33857600
2022-01-04,49.285557,50.027529,48.831100,49.684368,45681400
2022-01-05,49.962616,52.095788,49.925520,50.259405,59109300
2022-01-06,50.092461,50.704588,49.545255,50.648943,35757900
2022-01-07,49.563801,50.444895,49.294838,50.259401,30717200
...,...,...,...,...,...
2026-07-27,91.669998,94.980003,86.940002,92.459999,133247300
2026-07-28,86.300003,88.279999,83.099998,86.620003,154346700
2026-07-29,81.879997,88.470001,81.790001,86.570000,152394600


In [35]:
fig = go.Figure()

fig.add_trace(go.Candlestick(open=intel['Open'],close=intel['Close'],high=intel['High'],low=intel['Low'],x=intel.index))
fig.update_layout(title='Intel Candlestick Chart',
                  template='plotly_dark')
fig.show()

In [37]:
intel['SMA_20'] = intel['Close'].rolling(20).mean()
intel['SMA_50'] = intel['Close'].rolling(50).mean()
intel.dropna(inplace=True)
intel   

Price,Close,High,Low,Open,Volume,SMA_20,SMA_50
Date,,,,,,,
2022-03-15,41.876354,41.960462,41.184799,41.586648,34448900,43.780465,46.173907
2022-03-16,43.577206,43.577206,42.306242,42.334277,38885500,43.695890,46.058441
2022-03-17,44.053810,44.193991,42.736121,43.287495,28555000,43.644958,45.953806
2022-03-18,44.343525,44.483703,43.315538,43.754769,46237100,43.639351,45.841424
2022-03-21,44.287449,44.754715,43.988399,44.352866,32910900,43.749159,45.725324
...,...,...,...,...,...,...,...
2026-07-27,91.669998,94.980003,86.940002,92.459999,133247300,108.958500,114.694800
2026-07-28,86.300003,88.279999,83.099998,86.620003,154346700,106.687500,114.102200
2026-07-29,81.879997,88.470001,81.790001,86.570000,152394600,103.800000,113.564400


In [41]:
intel['Returns'] = intel['Close'].pct_change()
intel['Signal'] = np.where(intel['SMA_20'] > intel['SMA_50'], 1, 0)
intel['Signal_Returns'] = intel['Signal'].shift(1) * intel['Returns']
intel.dropna(inplace=True)
intel

Price,Close,High,Low,Open,Volume,SMA_20,SMA_50,Signal,Returns,Signal_Returns
Date,,,,,,,,,,
2022-03-16,43.577206,43.577206,42.306242,42.334277,38885500,43.695890,46.058441,0,0.040616,0.0
2022-03-17,44.053810,44.193991,42.736121,43.287495,28555000,43.644958,45.953806,0,0.010937,0.0
2022-03-18,44.343525,44.483703,43.315538,43.754769,46237100,43.639351,45.841424,0,0.006576,0.0
2022-03-21,44.287449,44.754715,43.988399,44.352866,32910900,43.749159,45.725324,0,-0.001265,-0.0
2022-03-22,45.221977,45.296741,44.296793,44.390245,27923300,43.922047,45.638488,0,0.021101,0.0
...,...,...,...,...,...,...,...,...,...,...
2026-07-27,91.669998,94.980003,86.940002,92.459999,133247300,108.958500,114.694800,0,-0.007041,-0.0
2026-07-28,86.300003,88.279999,83.099998,86.620003,154346700,106.687500,114.102200,0,-0.058580,-0.0
2026-07-29,81.879997,88.470001,81.790001,86.570000,152394600,103.800000,113.564400,0,-0.051217,-0.0


In [42]:
intel.groupby('Signal')['Signal_Returns'].mean()

Signal
0    0.000481
1    0.000111
Name: Signal_Returns, dtype: float64

In [43]:
intel['Equity_Curve'] = (1 + intel['Signal_Returns']).cumprod()
intel['Buy_Hold'] = (1 + intel['Returns']).cumprod()
intel

Price,Close,High,Low,Open,Volume,SMA_20,SMA_50,Signal,Returns,Signal_Returns,Equity_Curve,Buy_Hold
Date,,,,,,,,,,,,
2022-03-16,43.577206,43.577206,42.306242,42.334277,38885500,43.695890,46.058441,0,0.040616,0.0,1.000000,1.040616
2022-03-17,44.053810,44.193991,42.736121,43.287495,28555000,43.644958,45.953806,0,0.010937,0.0,1.000000,1.051997
2022-03-18,44.343525,44.483703,43.315538,43.754769,46237100,43.639351,45.841424,0,0.006576,0.0,1.000000,1.058916
2022-03-21,44.287449,44.754715,43.988399,44.352866,32910900,43.749159,45.725324,0,-0.001265,-0.0,1.000000,1.057577
2022-03-22,45.221977,45.296741,44.296793,44.390245,27923300,43.922047,45.638488,0,0.021101,0.0,1.000000,1.079893
...,...,...,...,...,...,...,...,...,...,...,...,...
2026-07-27,91.669998,94.980003,86.940002,92.459999,133247300,108.958500,114.694800,0,-0.007041,-0.0,0.849174,2.189063
2026-07-28,86.300003,88.279999,83.099998,86.620003,154346700,106.687500,114.102200,0,-0.058580,-0.0,0.849174,2.060829
2026-07-29,81.879997,88.470001,81.790001,86.570000,152394600,103.800000,113.564400,0,-0.051217,-0.0,0.849174,1.955280


In [44]:
fig = go.Figure()

fig.add_trace(go.Scatter(x=intel.index,y=intel['Equity_Curve'],name='Equity Curve',mode='lines'))
fig.add_trace(go.Scatter(x=intel.index,y=intel['Buy_Hold'],name='Buy & Hold',mode='lines'))
fig.update_layout(title='Intel Strategy vs Buy & Hold',
                  template='plotly_dark')
fig.show()

In [45]:
print(f'Sharpe ratio of strategy: {sharpe(intel['Signal_Returns'])}')
print(f'Sharpe ratio of holding: {sharpe(intel['Returns'])}')

Sharpe ratio of strategy: 0.15069627821770037
Sharpe ratio of holding: 0.5917798007159021


# Project Conclusion: SMA Crossover Strategy Analysis

## What Was Tested
A Simple Moving Average (SMA) crossover strategy was backtested 
on two stocks with very different characteristics:
- **AAPL (2025–2026)** — strong, consistent uptrend
- **INTC (2022–2026)** — choppy, declining market with one sharp spike

Two crossover variations were tested:
- Slow: SMA 20/50
- Fast: SMA 10/30

Performance was evaluated using Sharpe ratio and equity curves 
compared against a buy & hold benchmark.

---

## Results

| | AAPL SMA 20/50 | AAPL SMA 10/30 | AAPL Buy & Hold |
|---|---|---|---|
| Sharpe | 1.30 | 1.19 | 1.32 |

| | INTC SMA 20/50 | INTC Buy & Hold |
|---|---|---|
| Sharpe | 0.15 | 0.59 |

---

## Conclusions

**1. SMA crossover underperforms in strong uptrends**
On AAPL, buy & hold outperformed both strategies on a 
risk-adjusted basis. The strategy exits positions during 
pullbacks and re-enters too late, missing recovery gains.

**2. SMA crossover provides downside protection in choppy markets**
On INTC, the strategy avoided the majority of the 2022–2025 
decline by sitting in cash during extended death cross periods. 
The equity curve was significantly smoother than buy & hold.

**3. Faster crossover ≠ better results**
SMA 10/30 tracked buy & hold more closely on AAPL but produced 
a lower Sharpe than SMA 20/50, suggesting more signals 
introduced more noise rather than more edge.

---

## Key Learnings
- Signal quality matters more than signal frequency
- A strategy must be evaluated in the right market conditions
- Sharpe ratio alone can be misleading — always read the equity curve
- Buy & hold is a surprisingly strong benchmark to beat

---

## Next Steps
- Test mean reversion strategy on a choppy stock
- Add maximum drawdown metric to evaluation framework
- Test across more stocks to validate findings